modules/linear_boltzmann_solvers/c_discrete_ordinates_adjoint_solver/lbs_adj_solver.cc

```diff
@@ -160,38 +106,86 @@ DiscreteOrdinatesAdjointSolver::Execute()
       for (int m = 0; m < num_moments_; ++m)
       {
         const auto& ell = m_to_ell_em_map[m].ell;
-
-        size_t dof_map_g0 = cell_view.MapDOF(i, m, 0); // unknown map
+        const auto dof_map = cell_view.MapDOF(i, m, 0);
 
         for (int g : set_group_numbers)
-          phi_old_local_[dof_map_g0 + g] *= pow(-1.0, ell);
+          phi_old_local_[dof_map + g] *= pow(-1.0, ell);
       } // for moment
     }   // node i
   }     // for cell
 
   UpdateFieldFunctions();
 }
 
-size_t
-DiscreteOrdinatesAdjointSolver::AddResponseFunction(const std::string& qoi_name,
-                                                    std::shared_ptr<LogicalVolume> logical_volume,
-                                                    std::shared_ptr<ResponseFunction> function)
+double
+DiscreteOrdinatesAdjointSolver::ComputeInnerProduct()
 {
-  // Make the designation
-  ResponseFunctionDesignation qoi_designation(qoi_name, std::move(logical_volume), function);
-  // Make empty subscriber list (will be populated during initialize)
-  std::vector<size_t> cell_rf_subscriptions;
+  double local_integral = 0.0;
 
-  response_functions_.emplace_back(qoi_designation, cell_rf_subscriptions);
+  // Material sources
+  for (const auto& cell : grid_ptr_->local_cells)
+  {
+    if (matid_to_src_map_.count(cell.material_id_) == 0) continue; // Skip if no src
 
-  return response_functions_.size() - 1;
+    const auto& transport_view = cell_transport_views_[cell.local_id_];
+    const auto& source = matid_to_src_map_[cell.material_id_];
+    const auto& fe_values = unit_cell_matrices_[cell.local_id_];
+
+    for (const auto& group : groups_)
+    {
+      const int g = group.id_;
+      const double Q = source->source_value_g_[g];
+
+      if (Q > 0.0)
+      {
+        const int num_nodes = transport_view.NumNodes();
+        for (int i = 0; i < num_nodes; ++i)
+        {
+          const auto dof_map = transport_view.MapDOF(i, 0, g);
+          const auto& phi = phi_old_local_[dof_map];
+
+          local_integral += Q * phi * fe_values.Vi_vectors[i];
```

Comment:
I still get tripped on this. Vi_vectors is the volume basis function i. Correct? Pffff. since [i] is  used, the 'i' in 'Vi' is actually confusing. And one could be using Vi_vectors[j] as well. I know it's a small detail but it trips me constantly these days, so I feel that as some point, later, we may want to rethink some names.

Reply:
It trips me up too, and I'm not a fan of the naming either. That's why I created issue #73 a few PRs ago when you brought this up.